src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Linux.cs

```diff
@@ -22,5 +35,233 @@ private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[]
+#pragma warning disable CA2012 // Use ValueTasks correctly
+                DownloadOcsp();
+#pragma warning restore CA2012 // Use ValueTasks correctly
+            }
+
+            return ValueTask.FromResult(_ocspResponse);
+        }
+
+        private ValueTask<byte[]?> DownloadOcsp()
+        {
+            Task<byte[]?>? pending = _pendingDownload;
+
+            if (pending is not null && !pending.IsFaulted)
+            {
+                return new ValueTask<byte[]?>(pending);
+            }
+
+            if (_ocspUrls is null && _ca is not null)
+            {
+                foreach (X509Extension ext in Certificate.Extensions)
+                {
+                    if (ext is X509AuthorityInformationAccessExtension aia)
+                    {
+                        foreach (string entry in aia.EnumerateOcspUris())
+                        {
+                            if (Uri.TryCreate(entry, UriKind.Absolute, out Uri? uri))
+                            {
+                                if (uri.Scheme == UriScheme.Http)
+                                {
+                                    (_ocspUrls ??= new List<string>()).Add(entry);
+                                }
+                            }
+                        }
+
+                        break;
+                    }
+                }
+            }
+
+            if (_ocspUrls is null)
+            {
+                _ocspExpiration = _nextDownload = DateTimeOffset.MaxValue;
+                return new ValueTask<byte[]?>((byte[]?)null);
+            }
+
+            lock (SslContexts)
+            {
+                pending = _pendingDownload;
+
+                if (pending is null || pending.IsFaulted)
+                {
+                    _pendingDownload = pending = FetchOcsp();
+                }
+            }
+
+            return new ValueTask<byte[]?>(pending);
+        }
+
+        private async Task<byte[]?> FetchOcsp()
+        {
+            X509Certificate2? caCert = _ca;
+            Debug.Assert(_ocspUrls is not null);
+            Debug.Assert(_ocspUrls.Count > 0);
+            Debug.Assert(caCert is not null);
+
+            IntPtr subject = Certificate.Handle;
+            IntPtr issuer = caCert.Handle;
+
+            using (SafeOcspRequestHandle ocspRequest = Interop.Crypto.X509BuildOcspRequest(subject, issuer))
+            {
+                byte[] rentedBytes = ArrayPool<byte>.Shared.Rent(Interop.Crypto.GetOcspRequestDerSize(ocspRequest));
+                int encodingSize = Interop.Crypto.EncodeOcspRequest(ocspRequest, rentedBytes);
+                ArraySegment<byte> encoded = new ArraySegment<byte>(rentedBytes, 0, encodingSize);
+
+                ArraySegment<char> rentedChars = Base64UrlEncoding.RentEncode(encoded);
+                byte[]? ret = null;
+
+                for (int i = 0; i < _ocspUrls.Count; i++)
+                {
+                    string url = MakeUrl(_ocspUrls[i], rentedChars);
```

Comment:
Do we need to check the scheme here? Is there any risk of recursion like we saw with AIA fetching?

1. I need to establish TLS to example.org.
2. I want to fetch OCSP for a site to https://example.org/ since I am connecting to example.org.
3. OCSP responder is at https://example.org/ocsp
4. I need to establish TLS to example.org to check OCSP.
5. I want to fetch OCSP for a site to https://example.org/ since I am connecting to example.org.
6. etc.